Specify lang attribute in generated app layout

edwardloveall opened the pull request:
### Summary

This adds the lang attribute to the html tag in the generated application.html.erb layout. The W3C [states][1]:

> Always use a language attribute on the html tag to declare the default language of the text in the page.

Rails exposes the code for the current language with the `locale` helper. This seemed like the obvious choice to set the language properly and automatically.

### Why is this a problem?

First, declaring what language a page is written in is a good web practice.

Second, some browsers such as Google Chrome will ask if a site needs to be translated if it detects there are many non-native words on the page. It's possible to confuse this translation detector if, for
example, you have a quirky company name on a page with little other text like a promo page. Adding the lang attribute puts a stop to this unhelpful suggestion by the browser.

[1]:
https://www.w3.org/International/questions/qa-html-language-declarations

pixeltrix commented on railties/CHANGELOG.md:
```diff
@@ -1 +1,8 @@
+    W3C recommends the use of the lang attribute on the html tag. This uses
+    the locale view helper to specify the current app's language.
+
+    *Edward Loveall*
```
This changelog entry needs to be above the `Please check …` line

edwardloveall replied:
Thanks for the clarification, I was a little confused exactly how the CHANGELOG worked here. Fixed and repushed.

pixeltrix commented:
Not sure what our recommended pattern would be - dynamically outputting the locale or using locale specific templates, e.g. `application.en.html.erb`. However I'm in agreement with the sentiment behind the PR - [this commit][1] has some relevant links.

[1]: https://github.com/alphagov/e-petitions/commit/5e22a17ac8b3a235260a7e650e3ec604a6277607

pixeltrix commented on railties/lib/rails/generators/rails/app/templates/app/views/layouts/application.html.erb.tt:
```diff
@@ -1,5 +1,5 @@
 <!DOCTYPE html>
-<html>
+<html lang="<%%= locale %>">
```
The `locale` method appears to be undocumented - I assume it's the delegation to the lookup context in `ActionView::Base`. It's probably okay but don't want to accidentally commit us to supporting this method.

gmcgibbon replied:
`I18n.default_locale` seems like a valid choice but if we're interpolating that into the template on all newly generated apps I'd expect it to be first class. Similar to how you can do `t` instead of `I18n.t` in a view, can we make `default_locale` accessible in the view and use that?

edwardloveall replied:
Thinking about this a little more, does `locale` make more sense than `default_locale`? If the user changes  the locale (like [this](http://guides.rubyonrails.org/i18n.html#managing-the-locale-across-requests)) `default_locale` will still return `en`, but `locale` change along with the user.

edwardloveall commented on railties/lib/rails/generators/rails/app/templates/app/views/layouts/application.html.erb.tt:
```diff
@@ -1,5 +1,5 @@
 <!DOCTYPE html>
-<html>
+<html lang="<%%= locale %>">
```
That's pretty interesting that it's not documented in the API. I found it in the [internationalization guide](http://guides.rubyonrails.org/i18n.html#the-public-i18n-api) which is not shy about mentioning it.

pixeltrix replied:
I think those are attribute setters/getters on `I18n` though - have you tried just `default_locale` in a view?

edwardloveall replied:
It's undefined in a view. You have to do `I18n.default_locale`. I think I'd still prefer `locale`, but `I18n.default_locale ` is hardly a bad solution.

edwardloveall commented:
@pixeltrix is there anything else I should do to get this merged or looked at further?

gmcgibbon reviewed (commented):
Still relevant in Rails 6.

rafaelfranca commented on railties/CHANGELOG.md:
```diff
@@ -97,5 +104,4 @@
 
     *Jan Krutisch*
 
-
```
This line should stay 😄

edwardloveall commented:
Ugh, I'm having some git issues, obviously. Please ignore the ridiculous amount of changes until then

edwardloveall committed:
Specify lang attribute in generated app layout

This adds the lang attribute to the html tag in the generated
application.html.erb layout. The W3C [states][1]:

> Always use a language attribute on the html tag to declare the default
> language of the text in the page.

Rails exposes the code for the current language with the `locale`
helper. This seemed like the obviuous choice to set the language
properly and automatically.

Why is this a problem?

First, declaring what language a page is written in is a good web
practice.

Second, some browsers such as Google Chrome will ask if a site needs to
be translated if it detects there are many non-native words on the
page. It's possible to confuse this translation detector if, for
example, you have a quirky company name on a page with little other text
like a promo page. Adding the lang attribute puts a stop to this
unhelpful suggestion by the browser.

[1]:
https://www.w3.org/International/questions/qa-html-language-declarations
